### PR TITLE
Add spot instances configuration

### PR DESCRIPTION
**Pull request by MarcelMue**

### CODE REVIEW

**Comment by teemow** [comment]
LGTM. AWS has chosen a different path in terms of configuring this. In the MixedInstancePolicy you can set on-demand and spot-instance parameters. But I think it is fine to add a struct for the spot instances configuration only. This can then be mapped in the provider specific operators accordingly.

**Comment by MarcelMue** [comment]
We can change this later on easily. Let's see later if this is sufficient.

**Comment by teemow** [comment]
We also need to update the status:

 * Which alike instance types are allowed?
 * How many spot instances are used in the cluster?
 * How many of which instance type is used? (not sure if we really need this now - we should instead have a representation of the instances in a node pool in the future)

File: docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml [hunk 1]
```diff
@@ -83,6 +83,26 @@ spec:
...
                       type: string
                     type: array
+                  instanceDistribution:
+                    description: |
+                      Attributes defining the instance distribution in a node pool.
+                    properties:
+                      onDemandBaseCapacity:
+                        default: 0
+                        description: |
+                          Base capacity of on demand machines.
```
Comment (marians) [inline docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml]:
```suggestion
                          Base capacity of on-demand machines.
```

Comment (marians) [inline docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml]:
```suggestion
                          Base capacity of on-demand instances to use for worker nodes.
```

File: docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml [hunk 2]
```diff
@@ -83,6 +83,26 @@ spec:
...
+                        format: int32
+                        minimum: 0
+                        type: integer
+                      onDemandPercentageAboveBaseCapacity:
+                        default: 100
+                        description: |
+                          Percentage of on demand instances above the base capacity.
```
Comment (marians) [inline docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml]:
```suggestion
                          Percentage of on-demand instances to use for worker nodes, for instances exceeding onDemandBaseCapacity.
```

File: docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml [hunk 3]
```diff
@@ -91,6 +111,33 @@ spec:
                         description: |
                           AWS EC2 instance type name to use for the worker nodes in this node pool.
                         type: string
+                      useAlikeInstanceTypes:
+                        description: |
+                          If true, instances alike the instance_type will be used.
```
Comment (marians) [inline docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml]:
```suggestion
                          If true, certain instance types with specs similar to .spec.provider.worker.instanceType will be used.
```

**Comment by marians** [comment]
Kudos for also adding the `.status` section to the validation schema and thus publishing docs for this!

File: docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml (hunk 1 above)
Comment (marians) [reply docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml]: Sorry, my change suggestion is in the wrong place. Will move.

File: docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml (hunk 2 above)
Comment (marians) [reply docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml]: Sorry, my change suggestion is in the wrong place. Will move.

File: docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml (hunk 3 above)
Comment (marians) [reply docs/crd/infrastructure.giantswarm.io_awsmachinedeployment.yaml]: Sorry, my change suggestion is in the wrong place. Will move.

File: pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types.go [hunk 4]
```diff
@@ -99,6 +99,26 @@ spec:
...
                       type: string
                     type: array
+                  instanceDistribution:
+                    description: |
+                      Attributes defining the instance distribution in a node pool.
+                    properties:
+                      onDemandBaseCapacity:
+                        default: 0
+                        description: |
+                          Base capacity of on demand machines.
```
Comment (marians) [inline pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types.go]:
```suggestion
                          Base capacity of on-demand instances to use for worker nodes.
```

File: pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types.go [hunk 5]
```diff
@@ -99,6 +99,26 @@ spec:
...
+                        format: int32
+                        minimum: 0
+                        type: integer
+                      onDemandPercentageAboveBaseCapacity:
+                        default: 100
+                        description: |
+                          Percentage of on demand instances above the base capacity.
```
Comment (marians) [inline pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types.go]:
```suggestion
                          Percentage of on-demand instances to use for worker nodes, for instances exceeding onDemandBaseCapacity.
```

File: pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types.go [hunk 6]
```diff
@@ -108,6 +128,33 @@ spec:
                         description: |
                           AWS EC2 instance type name to use for the worker nodes in this node pool.
                         type: string
+                      useAlikeInstanceTypes:
+                        description: |
+                          If true, instances alike the instance_type will be used.
```
Comment (marians) [inline pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types.go]:
```suggestion
                          If true, certain instance types with specs similar to instanceType will be used.
```

File: pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types_test.go [hunk 7]
```diff
@@ -92,8 +92,24 @@ func newAWSMachineDeploymentExampleCR() *AWSMachineDeployment {
...
+		Provider: AWSMachineDeploymentStatusProvider{
+			Worker: AWSMachineDeploymentStatusProviderWorker{
+				InstanceTypes: []string{
+					"m4.xlarge",
+					"m5.xlarge",
+				},
+				SpotInstances: 39,
 			},
 		},
 	}
```
Comment (marians) [inline pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types_test.go]: I am thinking that a realistic example should have even more Status attributes defined. Can we add that as part of this PR, or should we make that separate?

Comment (MarcelMue) [reply pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types_test.go]: This is a complete example of all Status attributes of this CR - what others did you have in mind? 

This matches pretty much what we see in clusters currently.

Comment (marians) [reply pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types_test.go]: I was expecting the desired and actual number of instances, coming from the cluster-autoscaler. Is that in a different CR then? Anyway, comment solved.

Comment (MarcelMue) [reply pkg/apis/infrastructure/v1alpha2/aws_machine_deployment_types_test.go]: Yes that is different CR.

**Pull request merged**